operationsgateway_api/src/records/record.py

```diff
@@ -386,3 +413,264 @@ async def convert_search_ranges(date_range, shotnum_range):
+            if "metadata" in channel_value and "x_units" in channel_value["metadata"]:
+                x_units = channel_value["metadata"]["x_units"]
+            else:
+                x_units = None
+
+            waveform = Waveform.get_waveform(waveform_path)
+            return WaveformVariable(waveform, x_units=x_units)
+        else:
+            return channel_value["data"]
+
+    @staticmethod
+    def _parse_function_results(
+        record: dict,
+        original_image: bool,
+        lower_level: int,
+        upper_level: int,
+        colourmap_name: str,
+        function_name: str,
+        result: "np.ndarray | WaveformVariable | np.float64",
+        return_thumbnails: bool = True,
+    ) -> None:
+        """
+        Parses the numerical `result` and modifies `record` in place to contain
+        the data in the expected format for the type of the `result`.
+        """
+        Record._ensure_channels(record)
+
+        if isinstance(result, np.ndarray):
+            channel = Record._parse_image_result(
+                result=result,
+                original_image=original_image,
+                lower_level=lower_level,
+                upper_level=upper_level,
+                colourmap_name=colourmap_name,
+                return_thumbnails=return_thumbnails,
+            )
+
+        elif isinstance(result, WaveformVariable):
+            metadata = {"channel_dtype": "waveform"}
+            if result.x_units is not None:
+                metadata["x_units"] = result.x_units
+
+            if return_thumbnails:
+                waveform = result.to_waveform()
+                # Creating thumbnail takes ~ 0.05 s per waveform
+                waveform.create_thumbnail()
+                channel = {"thumbnail": waveform.thumbnail, "metadata": metadata}
+            else:
+                waveform_model = result.to_waveform_model()
+                channel = {"data": waveform_model, "metadata": metadata}
+
+        else:
+            metadata = {"channel_dtype": "scalar"}
+            # Cannot return np.float64 as it can't be cast to JSON
+            channel = {"data": float(result), "metadata": metadata}
+
+        record["channels"][function_name] = channel
+
+    @staticmethod
+    def _parse_image_result(
+        result: np.ndarray,
+        original_image: bool,
+        lower_level: int,
+        upper_level: int,
+        colourmap_name: str,
+        return_thumbnails: bool,
+    ) -> dict:
+        """Parses a numpy ndarray and returns image bytes, either for a thumbnail or
+        full image.
+        """
+        # We do not track the bit depth of inputs, so keep maximum depth to
+        # avoid losing information
+        bits_per_pixel = 16
+        if return_thumbnails:
+            metadata = {
+                "channel_dtype": "image",
+                "x_pixel_size": result.shape[1],
+                "y_pixel_size": result.shape[0],
+            }
+            step_size = max(*result.shape) // 50
```

Comment:
Yeah I likely used the default thumbnail size of 50 and never got round to generalising it to the actual setting.

It's taking the larger of width and height, then doing floor division to get the number of pixels in the original image that need to be mapped to one of the 50 output pixels. Then just use python slicing to downsample our original image. It's far more quick and dirty than using PIL like we do in the actual thumbnail script. Could do that, would probably introduce overhead but it might be small compared to getting the data from echo, haven't tested it (this was just the simplest way of getting a small image out).